frontend/src/utils/CachedObservable.ts

```diff
@@ -0,0 +1,129 @@
+import {
+  defer,
+  firstValueFrom,
+  Observable,
+  ObservableNotification,
+  OperatorFunction,
+  Subject,
+} from 'rxjs';
+import {
+  delayWhen,
+  dematerialize,
+  filter,
+  materialize,
+  repeatWhen,
+  shareReplay,
+  startWith,
+} from 'rxjs/operators';
+import { CancellationToken } from './CancellationToken';
+import withCancellation from './withCancellation';
+
+/**
+ * LoadingNotification like {@link ObservableNotification}.
+ */
+export interface LoadingNotification {
+  kind: 'L';
+}
+
+/**
+ * Extends {@link ObservableNotification} with {@link LoadingNotification}.
+ */
+export type ExtendedObservableNotification<T> =
+  | ObservableNotification<T>
+  | LoadingNotification;
+
+/**
+ * Returns true if param is {@link ObservableNotification}.
+ */
+export function isObservableNotification<T>(
+  state: ExtendedObservableNotification<T>
+): state is ObservableNotification<T> {
+  return state.kind === 'N' || state.kind === 'E' || state.kind === 'C';
+}
+
+/**
+ * TODO Change Comment since implementation has changed
+ * Observable with a cache.
+ * The source observable is only subscribed to after this observable
+ * is subscribed to (lazy evaluation).
+ * The cached value is the first value emitted from the given source observable.
+ * If an error happens, the error is forwarded to all active subscribers.
+ * The source action is automatically restarted after an error when a new
+ * subscription is made.
+ */
+export default abstract class CachedObservable<IN, OUT> {
```

Comment:
Geeneric argument should be prefixed with `T` and by in Pascal case. So the type arguments have to be names `TIn` and `TOut`

Reply:
Pascal: Yes
Prefix: [No](https://www.typescriptlang.org/docs/handbook/2/generics.html)
Will change it to `In` and `Out`